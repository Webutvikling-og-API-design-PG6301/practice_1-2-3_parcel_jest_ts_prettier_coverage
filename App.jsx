import React, { useState, useContext } from "react";
import { Routes, Route, Link, useNavigate } from "react-router-dom";
import { randomQuestion, isCorrectAnswer } from "./questions";

export const QuestionContext = React.createContext({ randomQuestion });

export const Home = ({ isRightAnswer, isQuestionAnswered }) => {
  const navigate = useNavigate();

  return (
    <div>
      <h1>This is the home page</h1>
      <h1 data-testid={"status"}>
        You have answered {isRightAnswer} / {isQuestionAnswered}
      </h1>
      <Link to="/question">
        <button>New Quiz</button>
      </Link>
    </div>
  );
};

export const Questions = ({ setIsRightAnswer, setIsQuestionAnswered }) => {
  const navigate = useNavigate();

  const handleAnswer = (answer) => {
    setIsQuestionAnswered((prev) => prev + 1);
    if (isCorrectAnswer(question, answer)) {
      setIsRightAnswer((prev) => prev + 1);
      navigate("/answer/correct");
    } else {
      navigate("/answer/wrong");
    }
  };

  const { randomQuestion } = useContext(QuestionContext);
  const [question] = useState(randomQuestion());

  return (
    <div>
      <h1>{question.question}</h1>
      {Object.keys(question.answers)
        .filter((value) => question.answers[value])
        .map((answer, idx) => {
          return (
            <div key={idx} data-testid={answer}>
              <button onClick={() => handleAnswer(answer)}>
                {question.answers[answer]}
              </button>
            </div>
          );
        })}
    </div>
  );
};

export const Answers = () => {
  return (
    <>
      <Routes>
        <Route path={"correct"} element={<h1>Correct answer</h1>} />
        <Route path={"wrong"} element={<h1>Wrong answer</h1>} />
      </Routes>
      <div>
        <Link to="/">
          <button>Whats my score?</button>
        </Link>
        <Link to="/question">
          <button>New Quiz</button>
        </Link>
      </div>
    </>
  );
};

const App = () => {
  const [isRightAnswer, setIsRightAnswer] = useState(0);
  const [isQuestionAnswered, setIsQuestionAnswered] = useState(0);
  return (
    <div>
      <h1>Welcome to quizmaster</h1>
      <Routes>
        <Route
          path={"/"}
          element={
            <Home
              isRightAnswer={isRightAnswer}
              isQuestionAnswered={isQuestionAnswered}
            />
          }
        />
        <Route
          path={"/question"}
          element={
            <Questions
              setIsRightAnswer={setIsRightAnswer}
              setIsQuestionAnswered={setIsQuestionAnswered}
            />
          }
        />
        <Route path={"/answer/*"} element={<Answers />} />
      </Routes>
    </div>
  );
};

export default App;
