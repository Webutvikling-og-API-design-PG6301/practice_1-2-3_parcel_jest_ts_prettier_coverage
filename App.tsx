import React, { useState, useContext, FC } from "react";
import { Routes, Route, Link, useNavigate } from "react-router-dom";
import { randomQuestion, isCorrectAnswer } from "./questions";

export const QuestionContext = React.createContext({ randomQuestion });

interface Props {
  isQuestionAnswered: number;
  isRightAnswer: number;
}

interface Isetter {
  setIsRightAnswer: (isRightAnswer: Isetter) => void;
  setIsQuestionAnswered: (isQuestionAnswered: Isetter) => void;
}

export const Home = ({ isRightAnswer, isQuestionAnswered }: Props) => {
  const navigate = useNavigate();

  return (
    <div>
      <h1>This is the home page</h1>
      <h1 data-testid={"status"}>
        You have answered {isRightAnswer} / {isQuestionAnswered}
      </h1>
      <Link to="/question">
        <button>New Quiz</button>
      </Link>
    </div>
  );
};

export const Questions = ({
  setIsRightAnswer,
  setIsQuestionAnswered,
}: Isetter) => {
  const navigate = useNavigate();

  const handleAnswer = (answer: string) => {
    setIsQuestionAnswered((pre) => pre + 1);
    if (isCorrectAnswer(question, answer)) {
      setIsRightAnswer((prev: number) => prev + 1);
      navigate("/answer/correct");
    } else {
      navigate("/answer/wrong");
    }
  };

  const { randomQuestion } = useContext(QuestionContext);
  const [question] = useState(randomQuestion());

  return (
    <div>
      <h1>{question.question}</h1>
      {Object.keys(question.answers)
        .filter((value) => question.answers[value])
        .map((answer, idx) => {
          return (
            <div key={idx} data-testid={answer}>
              <button onClick={() => handleAnswer(answer)}>
                {question.answers[answer]}
              </button>
            </div>
          );
        })}
    </div>
  );
};

export const Answers = () => {
  return (
    <>
      <Routes>
        <Route path={"correct"} element={<h1>Correct answer</h1>} />
        <Route path={"wrong"} element={<h1>Wrong answer</h1>} />
      </Routes>
      <div>
        <Link to="/">
          <button>Whats my score?</button>
        </Link>
        <Link to="/question">
          <button>New Quiz</button>
        </Link>
      </div>
    </>
  );
};

const App: FC = () => {
  const [isRightAnswer, setIsRightAnswer] = useState<number>(0);
  const [isQuestionAnswered, setIsQuestionAnswered] = useState<number>(0);
  return (
    <div>
      <h1>Welcome to quizmaster</h1>
      <Routes>
        <Route
          path={"/"}
          element={
            <Home
              isRightAnswer={isRightAnswer}
              isQuestionAnswered={isQuestionAnswered}
            />
          }
        />
        <Route
          path={"/question"}
          element={
            <Questions
              setIsRightAnswer={setIsRightAnswer}
              setIsQuestionAnswered={setIsQuestionAnswered}
            />
          }
        />
        <Route path={"/answer/*"} element={<Answers />} />
      </Routes>
    </div>
  );
};

export default App;
