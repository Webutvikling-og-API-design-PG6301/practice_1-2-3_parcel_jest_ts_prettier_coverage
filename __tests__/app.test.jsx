import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter } from "react-router-dom";
import App from "../App";
import { Questions, Home, QuestionContext } from "../App";
import { Simulate } from "react-dom/test-utils";

const question = {
  question: "Are you happy?",
  answers: {
    answer_a: "Yes",
    answer_b: "No",
    answer_c: "Maybe",
  },
  correct_answers: {
    answer_a_correct: "true",
    answer_b_correct: "false",
    answer_c_correct: "false",
  },
};

describe("master quiz", () => {
  it("Should show the answers snapshot", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Home isRightAnswer={5} isQuestionAnswered={22} />
      </MemoryRouter>,
      element
    );
    expect(
      element.querySelector("[data-testid=status]").innerHTML
    ).toMatchSnapshot();
  });

  it("should show questions", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter initialEntries={["/question"]}>
        <QuestionContext.Provider value={{ randomQuestion: () => question }}>
          <App />
        </QuestionContext.Provider>
      </MemoryRouter>,
      element
    );
    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should show correct answer", () => {
    const setIsQuestionAnswered = jest.fn();
    const setIsRightAnswer = jest.fn();

    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter initialEntries={["/question"]}>
        <QuestionContext.Provider value={{ randomQuestion: () => question }}>
          <Questions
            setIsRightAnswer={setIsRightAnswer}
            setIsQuestionAnswered={setIsQuestionAnswered}
          />
        </QuestionContext.Provider>
      </MemoryRouter>,
      element
    );

    Simulate.click(element.querySelector("[data-testid=answer_a] button"));
    expect(setIsQuestionAnswered).toBeCalled();
    expect(setIsRightAnswer).toBeCalled();
    expect(element.innerHTML).toMatchSnapshot();
  });
});
