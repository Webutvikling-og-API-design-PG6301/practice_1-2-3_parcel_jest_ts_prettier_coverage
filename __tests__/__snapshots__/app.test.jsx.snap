// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home page Should show the component snapshot 1`] = `"You have answered 5 / 22"`;
